client/connection_pool.go

```diff
@@ -210,54 +214,69 @@ func (p *connPool) healthCheckEvery(interval time.Duration, stutter time.Duratio
 	for {
 		p.RLock()
 		state := p.state
-		poolLen := int(p.poolLen)
 		p.RUnlock()
 		if state != stateOpen {
 			return
 		}
 
-		start := nowFn()
-		deadline := start.Add(interval + randStutter(p.healthCheckRand, stutter))
-		// As this is the only loop that removes connections it is safe to loop upwards
-		for i := 0; i < poolLen; i++ {
-			p.RLock()
-			client := p.pool[i].client
-			p.RUnlock()
-
-			healthy := true
-			if err := p.healthCheck(client, p.opts); err != nil {
-				healthy = false
-				log.Warnf("health check failed to %s: %v", p.host.Address(), err)
-			}
-
-			if !healthy {
-				// Swap with tail, decrement pool size, decrement "i" to try next in line
-				p.Lock()
-				if p.state != stateOpen {
-					p.Unlock()
-					return
-				}
-				unhealthy := p.pool[i]
-				p.pool[i], p.pool[p.poolLen-1] = p.pool[p.poolLen-1], p.pool[i]
-				p.pool = p.pool[:p.poolLen-1]
-				p.poolLen = int64(len(p.pool))
-				p.Unlock()
-				unhealthy.channel.Close()
+		var (
+			wg       sync.WaitGroup
+			start    = nowFn()
+			deadline = start.Add(interval + randStutter(p.healthCheckRand, stutter))
+		)
 
-				// Retry with the element we just swapped in
-				i--
-			}
+		p.RLock()
+		for i := int64(0); i < p.poolLen; i++ {
+			wg.Add(1)
+			go func(client rpc.TChanNode) {
+				defer wg.Done()
+
+				var (
+					attempts = healthCheckFall
+					failed   = 0
+					checkErr error
+				)
+				for j := 0; j < attempts; j++ {
+					if err := p.healthCheck(client, p.opts); err != nil {
+						checkErr = err
+						failed++
+						continue
```

Comment:
Should we add some throttling here perhaps? If the health check failures were due to cpu usage spikes, we'll probably need to wait for the cpu usage to subside before the next health check suceeds